lib/useOnyx.ts

```diff
@@ -0,0 +1,200 @@
+import {deepEqual} from 'fast-equals';
+import {useCallback, useEffect, useRef, useSyncExternalStore} from 'react';
+import Onyx from './Onyx';
+import type {CollectionKeyBase, KeyValueMapping, OnyxCollection, OnyxEntry, OnyxKey, Selector} from './types';
+import usePrevious from './usePrevious';
+
+type OnyxValue<TKey extends OnyxKey> = TKey extends CollectionKeyBase ? OnyxCollection<KeyValueMapping[TKey]> : OnyxEntry<KeyValueMapping[TKey]>;
+
+type UseOnyxOptions<TKey extends OnyxKey, TReturnData> = {
+    /**
+     * Determines if this key in this subscription is safe to be evicted.
+     */
+    canEvict?: boolean;
+
+    /**
+     * If set to false, then no data will be prefilled into the component.
+     */
+    initWithStoredValues?: boolean;
+
+    /**
+     * TODO: Check if we still need this flag and associated logic.
+     */
+    // allowStaleData?: boolean;
```

Comment:
@roryabraham @tgolen Could you explain me what this `allowStaleData` property do exactly in `withOnyx` HOC?

Reply:
I'll defer most of my knowledge about it to this code comment:

https://github.com/Expensify/react-native-onyx/blob/10b74b7fd8b68ba9280b14015d62d1c6cefe3026/lib/withOnyx.js#L66-L76

Reply:
My basic understanding is that it allows the component to be rendered immediately instead of the next tick due to cache optimization.

Reply:
Yeah I think this is still needed.

- A component can render immediately if the value for a certain onyx key has already been cached
- However, the component we want to render might misses to be rendered immediately if the value is currently being updated
- This flag enables to render the component with "stale" data (the data before the update will happen)

(Reading through that code makes me wonder why we can't update the cache directly once we merge new values, but thats probably a different topic)

Reply:
One side note: I still have to review the whole PR, maybe the usage of that prop become obsolete (let me double check 🔍 )

Reply:
@hannojg Please share if you have some basic example about how to test this flag with `withOnyx` (we have only a few usages in the app), so I can also compare here if the behavior is consistent with `useOnyx`.

Reply:
Push okay, so actually after reviewing the PR I feel like maybe this exact option is needed anymore actually!

However, I am wondering, what happens to the use case of `!Onyx.hasPendingMergeForKey(key)`:

https://github.com/Expensify/react-native-onyx/blob/10b74b7fd8b68ba9280b14015d62d1c6cefe3026/lib/withOnyx.js#L77-L79

It has been initially added here:

- https://github.com/Expensify/react-native-onyx/pull/303

I believe right now we are missing that behaviour and the `useOnyx` hook would show stale data, e.g. meaning this behaviour:

```js
Onyx.merge('report_123', newValue); 
Navigation.navigate('report_screen', 'report_123');

// In ReportScreen

const data = useOnyx('report_123')
```

data would first be `staleData` and then in a second render become `newValue`.

Right now we are explicitly not using the cache value if has `Onyx.hasPendingMergeForKey(key)`.
However we made it possible to allow for reading stale data by that flag. 

So I think both should be added back:

- The  `Onyx.hasPendingMergeForKey(key)` for reading from cache or nah
- The `allowStaleData` flag

Reply:
Thanks for the example @hannojg ❤️ I will try to reproduce here and do the adjustments if necessary